compiler/modules/front.syntax/token.bal

```diff
@@ -487,9 +487,90 @@ public readonly class SourceFile {
         return unpackPosition(pos);
     }
 
-    public function lineContent(Position pos) returns string {
-        int lineNum = self.lineColumn(pos)[0];
-        return scanLineToString(self.scannedLine(lineNum));
+    // range is expected to be the start of a token
+    public function lineContent((Position|d:Range) range) returns [string, string, string] {
```

Comment:
Move the real work of this out into a module-level function in token.bal.

I think it will simplify things if we have a function (in scan.bal) that takes a scanLine and returns a `string[]` one for each member of fragCodes, such that scanLineToString is just the concatenation of those strings.

Then you can just iterate over this array until you the total length is what you need.

You can get rid of advanceFragmentIndex and fragmentToString.

Reply:
fixed